### DISCUSSION TITLE
Steps to reproduce a similar setup for further experimentation

### DISCUSSION THREAD

**Discussion by mrchrisadams** [dos-group/vessim#225] [Q&A]
Hi there,

I just finished reading this paper about Vessim properly, as a prelude to the HotCarbon conf next week:

> ### Software-in-the-loop simulation for developing and testing carbon-aware applications
> 
> ### Abstract
> --------
> 
> The growing electricity demand of IT infrastructure has raised significant concerns about its carbon footprint. To mitigate the associated emissions of computing systems, current efforts therefore increasingly focus on aligning the power usage of software with the availability of clean energy. To operate, such carbon-aware applications require visibility and control over relevant metrics and configurations of the energy system. However, research and development of novel energy system abstraction layers and interfaces remain difficult due to the scarcity of available testing environments: Real testbeds are expensive to build and maintain, while existing simulation testbeds are unable to interact with real computing systems. 
>
> To provide a widely applicable approach for developing and testing carbon-aware software, we propose a method for integrating real applications into a simulated energy system through software-in-the-loop simulation. The integration offers an API for accessing the energy system, while continuously modeling the computing system's power demand within the simulation. Our system allows for the integration of physical as well as virtual compute nodes, and can help accelerate research on carbon-aware computing systems in the future.

Source: [Software‐in‐the‐loop simulation for developing and testing carbon‐aware applications](https://onlinelibrary.wiley.com/doi/10.1002/spe.3275)

This is such a cool paper!

I have some questions about reproducing a setup for experimentation in future, as I am a contributor to [Branch Magazine](https://branch.climateaction.tech/), and have some familiarity [with the Solar Protocol art project](https://github.com/alexnathanson/solar-protocol). So, I am curious about adopting some of this for other projects.

#### 1. How would I replicate this setup used in the paper for future experimentation?

I appreciate that the paper outlines different ways to setup a 'test lab', and it seems a slightly more lightweight setup than replicating what the Ecovisor used.

I also see there is _still_ some hwardware required - particulary for physical Rpi nodes. I see the [the Solar Protocol documentation has some instructions for recreating a hardware setup](https://docs.google.com/document/d/1hdcTf9xUmsjRPd3waJEkQf1Bjive8Z6RmyWv_p5n8Is/edit) - would, there be interest in doing the same for the computing systems below?

![spe3275-fig-0003-m](https://github.com/dos-group/vessim/assets/17906/179ff873-afc5-4b93-a98f-180e21e8772e)

#### 2. Roughly how much would a setup like this cost?

Pretty self explanatory. Raspberry Pis are quite cheap, but I know much less about sourcing the sensors and batteries. When I've spoken to [another person building a similar solar + Pi hosting project, Scott Web](https://scott.ee/project/solar-hosting-raspberry-pi/), he's mentioned figures in the low hundreds for example, and having some kind of bill of materials would be very useful.

Knowing this would really help for budgeting to experiment with the ideas in this paper.

Thanks again for sharing all this so openly - it's really exciting to see stuff build on top of then ideas in the Ecovisor paper 👍

**Comment by birnbaum** [comment]
Hi Chris, thanks a lot for reaching out!

> I also see there is _still_ some hwardware required

You don't need hardware! There is the possibility to integrate hardware into your experiments if you want to, but you can also do everything through software. In this case, you need to use software-based power measurements like Intel's [Running Average Power Limit](https://www.intel.com/content/www/us/en/developer/articles/technical/software-security-guidance/advisory-guidance/running-average-power-limit-energy-reporting.html) or [nvidia-smi](https://developer.nvidia.com/system-management-interface). Or you can simply define some mapping function for e.g. CPU usage to power demand, as we did for the virtual node in the paper.

> Roughly how much would a setup like this cost?

For our hardware setup we spent less than 100€: it's really just a RaspberryPi 4b with a power sensor (we used [this one](https://www.switchdoc.com/2019/10/new-product-cvsensor-usb-to-usb-current-voltage-sensor-i2c-ina219/) but I think it's not available any more). But this setup is just a proof of concept, I wouldn't necessarily recommend it for serious experiments. Our current setup for testing cloud applications with Vessim includes power-metered 16-core servers, some including GPUs, so then we're obviously talking a lot more.

If you want to get started I recommend a software-based setup! Let us know what you want to test and we might be able to help you with that :)

**Comment by mrchrisadams** [comment]
> Or you can simply define some mapping function for e.g. CPU usage to power demand, as we did for the virtual node in the paper.

Ah.. I think you're talking about the CPU power curves of the kind discussed below and published by SPEC, right? 

I know the Green Coding solutions folk are working on a project to make them easy to generate based on the kind of underlying processor:

https://github.com/green-coding-solutions/cloud-energy

You can see how it tracks against the SPEC model, and other models used in the public domain:

![link](https://github.com/green-coding-solutions/cloud-energy/raw/main/img/hp_synergy_480_Gen10_Plus.png)

This makes me think you could set up a fairly general "virtual node" on most cloud platforms to model changes if you at least have access to the utilisation of the CPU.

Thanks for the pointer about the [Pi power measurement tool](https://www.switchdoc.com/2019/10/new-product-cvsensor-usb-to-usb-current-voltage-sensor-i2c-ina219), too - I didn't know they existed til this paper.

**Reply by birnbaum** [reply]
exactly! You just need to take into account that any power model will stay an approximation. We do not know about power demand besides the CPU, VM consolidation, and also there are different methodologies for attributing the demand of physical machines to virtual ones.
